### PR TITLE
Avoid overriding explicitly set grpc dial options

### LINKED ISSUE #2940: OTLP GRPC Span Exporter clobbers grpc.DialOptions
### Description

The [otlptracegrpc.WithDialOption](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc#WithDialOption) option is documented as being appended to the other grpc dial options and as such taking precedence over other options. The code does not conform to this contract. In my particular case an option of  `otlptracegrpc.WithDialOption(grpc.WithInsecure())` passed to `otlptracegrpc.New` will result in a grpc client configuration that expects a TLS  server. 

### Environment

- OS: all
- Architecture: all
- Go Version: all
- opentelemetry-go version: 1.7.0

### Steps To Reproduce

1. Start a non-TLS OTLP server endpoint (otelcol will do)
2. Create an OTLP GRPC trace exporter with  `otlptracegrpc.WithDialOption(grpc.WithInsecure())`
3. Note that span exports timeout because the exporter is trying to connect over TLS

[Functioning repro code](https://github.com/mmcshane/otel-grpc-exporter-repro).

### Expected behavior

Expect the grpc dial option to work

### PR DESCRIPTION
**Pull request by mmcshane**
The contract of oteltracegrpc.WithDialOption is that the options
supplied become the tail end of the dial option slice that is eventually
supplied to grpc.Dial however in some cases the implementation violates
that contract and appends additional dial options at the end of said
slice.

This commit changes otlpconfig option processing to ensue that the
contract of oteltracegrpc.WithDialOption is upheld. That is, the options
supplied become the suffix of the slice used to dial.

Fixes #2940

### CODE REVIEW

**Review by hanyuancheung — APPROVED** [review]
lgtm!

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Support for go1.16. Support is now only for go1.17 and go1.18 (#2917)
 
+### Fixed
+
+-  Avoid overriding explicitly set gRPC dial options in `go.opentelemetry.io/otel/exporters/otlptrace` (#2942)
```
Comment (hanyuancheung) [inline CHANGELOG.md:21]:
```suggestion
- Avoid overriding explicitly set gRPC dial options in `go.opentelemetry.io/otel/exporters/otlptrace` (#2942)
```

**Review by dmathieu — APPROVED** [review]

File: exporters/otlp/otlptrace/internal/otlpconfig/export_test.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlpconfig
+
+// NewGenericOption exposes otlpconfig.newGenericOption for tests.
+var NewGenericOption = newGenericOption
```
Comment (MadVikingGod) [inline exporters/otlp/otlptrace/internal/otlpconfig/export_test.go:18]: This is a neat trick to make this type available to the tests.

**Review by MadVikingGod — COMMENTED** [review]
The only question I have is does the [comment of WithDialOptions](https://github.com/open-telemetry/opentelemetry-go/blob/4afe160b707d7e3623deaf013666c76832b78f01/exporters/otlp/otlptrace/otlptracegrpc/options.go#L136-L142) still make sense after this change?

**Comment by mmcshane** [comment]
> The only question I have is does the [comment of WithDialOptions](https://github.com/open-telemetry/opentelemetry-go/blob/4afe160b707d7e3623deaf013666c76832b78f01/exporters/otlp/otlptrace/otlptracegrpc/options.go#L136-L142) still make sense after this change?

The comment was and remains sensible, the problem is that it was not correct with respect to the implementation. This patch brings the code into alignment with the comment. I.e. the code didn't do what the comment says and now (I _think_) it does.

**Comment by dmathieu** [comment]
No need to squash the commits. It will be done by GitHub at merge time.

**Comment by MadVikingGod** [comment]
So I don't think this patch makes things any better.

Take using these two options together:

`otlptracegrpc.WithServiceConfig("foo")` and `otlptracegrpc.WithDialOptions(grpc.WithDefaultServiceConfig("bar"))`

With the status quo if you do:

```go
otlptracegrpc.WithServiceConfig("foo"),
otlptracegrpc.WithDialOptions(grpc.WithDefaultServiceConfig("bar")),

// You get a config with foo.
```

But with your changes if you do:

```go
otlptracegrpc.WithDialOptions(grpc.WithDefaultServiceConfig("bar")),
otlptracegrpc.WithServiceConfig("foo"),

// You get a config with bar.
```

Both cases don't really make sense, and I don't think one is particularly better than the other.  What I would rather see is that if we have compeating ways of adding the same config they play well together, so both overwrite the other.  If that isn't possible then the comments need to be extra clear that using one will overwrite the other. (`// This option has no effect if WithGRPCConn is used or if WithDialOptions is used.`)

**Comment by mmcshane** [comment]
Rework the overall approach as and when you see fit. The intention here is to only bring the behavior in line with the existing documentation.

I'm only one datapoint but my reading of the docs resulted in an understanding that a tie-breaker for conflicting options had indeed been chosen: ["The options here are appended to the internal grpc.DialOptions used so they will take precedence over any other internal grpc.DialOptions they might conflict with."](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc#WithDialOption). I don't really know the history of this package but that seems like a clear and intentional statement. When you say "the comments need to be extra clear that using one will overwrite the other," you're correct, but I _that is the case already_.

I coded with the expectation that this package would work as indicated by the docs. The fact that it doesn't is what led to this PR. The details are in the bug report and the reproduction.

**Comment by MadVikingGod** [comment]
It's clear that the wording there is ambiguous. To me 'other internal grpc.DialOptions` means the defaults, or options not exposed.  Where you take it to mean something that includes the other Options in that package.  My ask is to change the wording to be more explicit, so users that follow after you won't deal with the ambiguity you found here.

**Comment by MrAlias** [comment]
This looks to have stalled. Please re-open if you intend to pick this back up.

**Pull request closed without merging**